### PR TITLE
feat: implement new select menu components

### LINKED ISSUE #798: New select menu components
### Summary

Add new options to select menu

### What is the feature request for?

The core library

### The Problem

Add these new select menu components to library as discord released these recently:

- User selects (type 5)
- Role selects (type 6)
- Mentionable (user and role) selects (type 7)
- Channel selects (type 8)

### The Ideal Solution

Options to select user/role/channel in select component.

### The Current Solution

None

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by shiftinv**
## Summary

Adds new select menu components/items, with <sup><sub>almost</sub></sup> everything there is to it (component types, methods, decorators, ...).

This is a draft PR for several reasons:
- ~~It depends on #800, which is currently set as the base branch to get a clean diff.~~
- ~~It depends on #801, which is currently merged into this branch to avoid merge conflicts later on.~~
- ~~I haven't added backwards-compatible aliases yet, but that'll be easy enough once everything else is settled.~~
- ~~Depends on #814 for the `resolved` field, currently set as base branch.~~
- May want to consider fixing the `Select.values` issue with concurrent interactions.

Resolves #798.

## Checklist

- [x] If code changes were made, then they have been tested
    - [x] I have updated the documentation to reflect the changes
    - [x] I have formatted the code properly by running `task lint`
    - [x] I have type-checked the code by running `task pyright`
- [ ] This PR fixes an issue
- [x] This PR adds something new (e.g. new method or parameters)
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

**Comment by onerandomusername** [comment]
@shiftinv would you please resolve conflicts?

**Review by onerandomusername — COMMENTED** [review]
Mostly good, but values ends up empty when used with disnake.ui.ActionRow.rows_from_message:

```py
class PR803(commands.Cog):
    @commands.slash_command()
    async def select(self, inter: disnake.GuildCommandInteraction) -> None:
        comps = disnake.ui.RoleSelect(custom_id="pr:803")
        await inter.send(components=comps)

    @commands.Cog.listener()
    async def on_dropdown(self, inter: disnake.MessageInteraction):
        if inter.data.custom_id != "pr:803":
            return
        comp: disnake.BaseSelectMenu = inter.component  # type: ignore
        print(inter.data.values)
        rows = disnake.ui.ActionRow.rows_from_message(inter.message)
        select = rows[0][0]
        await inter.send(repr(select))
        print(select.values)
```


Additionally, I have not yet tested their compatibility with views.

File: disnake/components.py [hunk 1]
```diff
@@ -276,15 +295,19 @@ def to_dict(self) -> BaseSelectMenuPayload:
         return payload
 
 
-class SelectMenu(BaseSelectMenu):
+# TODO: make all string-select-menu-related changes backwards-compatible
```
[the remark this answers is not in the document]
Comment (shiftinv) [reply disnake/components.py]: https://github.com/DisnakeDev/disnake/pull/803/commits/4476a32eea32baf27e3c1141b6416a5327955d7f

File: disnake/interactions/message.py [hunk 2]
```diff
@@ -74,13 +79,33 @@ class MessageInteraction(Interaction):
 
     def __init__(self, *, data: MessageInteractionPayload, state: ConnectionState):
         super().__init__(data=data, state=state)
-        self.data: MessageInteractionData = MessageInteractionData(data=data["data"])
+        self.data: MessageInteractionData = MessageInteractionData(
+            data=data["data"], state=state, guild_id=self.guild_id
+        )
         self.message = Message(state=self._state, channel=self.channel, data=data["message"])
 
+    # TODO: perhaps keep `values` like before and add a `resolved_values` instead?
```
Comment (shiftinv) [inline disnake/interactions/message.py:87]: unsure about this

File: disnake/interactions/base.py [hunk 3]
```diff
@@ -1762,26 +1763,26 @@ def __repr__(self):
         )
 
     def get_with_type(
-        self, key: Snowflake, data_type: OptionType, default: T = None
+        self, key: Snowflake, data_type: Union[OptionType, ComponentType], default: T = None
     ) -> Union[Member, User, Role, InteractionChannel, Message, Attachment, T]:
-        if data_type is OptionType.mentionable:
+        if data_type is OptionType.mentionable or data_type is ComponentType.mentionable_select:
```
Comment (onerandomusername) [inline disnake/interactions/base.py:1768]: is this better?
```suggestion
        if data_type in (OptionType.mentionable, ComponentType.mentionable_select):
```

would also be added for all of the other checks below

File: disnake/state.py [hunk 4]
```diff
@@ -952,7 +963,7 @@ def parse_interaction_create(self, data: gateway.InteractionCreateEvent) -> None
             self.dispatch("message_interaction", interaction)
             if interaction.data.component_type is ComponentType.button:
                 self.dispatch("button_click", interaction)
-            elif interaction.data.component_type is ComponentType.select:
+            elif interaction.data.component_type in _SELECT_COMPONENT_TYPES:
```
Comment (onerandomusername) [inline disnake/state.py:966]: this makes this change breaking, because on_dropdown now takes a BaseSelectMenu instead of SelectMenu (and that documentation change should be made)

File: disnake/ui/action_row.py [hunk 5]
```diff
@@ -343,6 +353,217 @@ def add_select(
         )
         return self
 
+    add_select = add_string_select  # backwards compatibility
```
Comment (onerandomusername) [inline disnake/ui/action_row.py:356]: undoc or redoc add_select in api.rst

File: disnake/components.py [hunk 6]
```diff
@@ -307,18 +330,151 @@ class SelectMenu(BaseSelectMenu):
...
         return payload
 
 
+SelectMenu = StringSelectMenu  # backwards compatibility
+
+
+class UserSelectMenu(BaseSelectMenu):
+    """Represents a user select menu from the Discord Bot UI Kit.
+
+    .. versionadded:: 2.7
```
Comment (onerandomusername) [inline disnake/components.py:351]:
```suggestion
    """Represents a user select menu from the Discord Bot UI Kit.

    .. note::
        The user constructible and usable type to create a
        user select menu is :class:`disnake.ui.UserSelect`.
        
    .. versionadded:: 2.7
```

Comment (onerandomusername) [reply disnake/components.py:351]: repeat for the other three classes

File: disnake/interactions/message.py (hunk 2 above)
Comment (onerandomusername) [reply disnake/interactions/message.py:87]: I think we should make this change be non-breaking and use resolved_values.

File: disnake/ui/item.py [hunk 7]
```diff
@@ -81,7 +81,7 @@ class Item(WrappedComponent, Generic[V_co]):
     The current UI items supported are:
 
     - :class:`disnake.ui.Button`
-    - subtypes of :class:`disnake.ui.BaseSelect` (:class:`disnake.ui.Select`)
+    - subtypes of :class:`disnake.ui.BaseSelect` (:class:`disnake.ui.StringSelect`, :class:`disnake.ui.UserSelect`, :class:`disnake.ui.RoleSelect`, :class:`disnake.ui.MentionableSelect`, :class:`disnake.ui.ChannelSelect`)
```
Comment (onerandomusername) [inline disnake/ui/item.py:84]: thoughts on ordering the selects alphabetically?

File: disnake/ui/select/channel.py [hunk 8]
```diff
@@ -0,0 +1,215 @@
...
+    "channel_select",
+)
+
+
+class ChannelSelect(BaseSelect[ChannelSelectMenu, "InteractionChannel", V_co]):
+    """Represents a UI channel select menu.
+
+    This is usually represented as a drop down menu.
+
+    In order to get the selected items that the user has chosen, use :attr:`.values`.
```
Comment (onerandomusername) [inline disnake/ui/select/channel.py:30]: not part of this pr, but the way to fix this state between different callbacks is a contextvar.

File: disnake/interactions/message.py (hunk 2 above)
Comment (onerandomusername) [reply disnake/interactions/message.py:87]: Hmmm, I'm now unsure too because the new selects have `values` refer to the values chosen and they are the resolved objects.

File: docs/api.rst [hunk 9]
```diff
@@ -5956,6 +6017,11 @@ ActionRow
 
 .. autoclass:: disnake.ui.ActionRow
     :members:
+    :exclude-members: add_select
+
+    .. method:: add_select
+
+        Alias of :func:`~disnake.ui.ActionRow.add_string_select`.
```
Comment (onerandomusername) [inline docs/api.rst:6024]:
```suggestion
        .. deprecated:: 2.7

        Alias of :func:`~disnake.ui.ActionRow.add_string_select`.
```

File: changelog/800.feature.rst [hunk 10]
```diff
@@ -1 +1 @@
-Separate select menu types (:class:`SelectMenu`, :class:`ui.Select`) into base types (:class:`BaseSelectMenu`, :class:`ui.BaseSelect`).
+TODO: replace with contents of main changelog entry
```
Comment (onerandomusername) [inline changelog/800.feature.rst:1]: needs changelog

File: changelog/803.feature.rst [hunk 11]
```diff
@@ -0,0 +1,3 @@
+Add new select menu components.
+
+TODO: more text c:
```
[the remark this answers is not in the document]
Comment (onerandomusername) [reply changelog/803.feature.rst:3]: needs changelog

File: disnake/interactions/base.py (hunk 3 above)
Comment (shiftinv) [reply disnake/interactions/base.py:1768]: That would work with the current enum implementation, but break with the introduction of #556 since `in` also checks `__eq__` in addition to identity checks:
```py
from disnake import Enum

class OptionType(Enum):
    role = 8

class ComponentType(Enum):
    channel_select = 8

print(OptionType.role is ComponentType.channel_select)  # False
print(OptionType.role == ComponentType.channel_select)  # False
print(OptionType.role in (ComponentType.channel_select,))  # False
```

With the new enums (or just stdlib enums), and using `int, Enum` as the base, you get `False, True, True`.

Either we keep the identity checks here, or avoid `Union[OptionType, ComponentType]` by splitting `InteractionDataResolved` (since that is arguably a bit weird anyway).

File: disnake/state.py (hunk 4 above)
Comment (shiftinv) [reply disnake/state.py:966]: `on_dropdown` receives the `interaction` object, so this isn't a breaking change.
If the component was passed directly, I'd still argue it wouldn't be breaking, since getting an unexpected component type would require the bot to send such a component in the first place.

File: disnake/ui/item.py (hunk 7 above)
Comment (shiftinv) [reply disnake/ui/item.py:84]: I ordered them based on the component types (3, 5, 6, 7, 8), did the same thing for the components and items on the api reference page; I'd prefer it that way.

File: docs/api.rst (hunk 9 above)
Comment (shiftinv) [reply docs/api.rst:6024]: Should the old methods/types really be deprecated? As of right now everything else is just aliased with no warning, and I'm personally fine with that (could change it in 3.0, perhaps).
I feel like the general perception is that the new selects are specialized forms of the "classic" select, instead of completely independent types, so there's still this notion of "a select menu" being the classic string select menu. Maybe that's just me tho.

File: disnake/components.py (hunk 6 above)
Comment (shiftinv) [reply disnake/components.py]: https://github.com/DisnakeDev/disnake/pull/803/commits/9d653c1502344c03a31a922798d2a696c8dbcdea

File: disnake/interactions/base.py (hunk 3 above)
Comment (onerandomusername) [reply disnake/interactions/base.py:1768]: ah, I see

File: disnake/state.py (hunk 4 above)
Comment (onerandomusername) [reply disnake/state.py:966]: right

File: disnake/interactions/message.py (hunk 2 above)
Comment (shiftinv) [reply disnake/interactions/message.py]: Yeah. For now I've decided to have a separate field, since the typing in existing bot code that uses `MessageInteraction.values` for string selects would otherwise be completely borked.
This way there's also no confusion between `inter.values` and `inter.data.values`.
https://github.com/DisnakeDev/disnake/pull/803/commits/c7ca5ea4db6db971c1cee6afd7d8464983f22d0a

File: changelog/800.feature.rst (hunk 10 above)
[the remark this answers is not in the document]
Comment (shiftinv) [reply changelog/800.feature.rst]: https://github.com/DisnakeDev/disnake/pull/803/commits/f59f9175b98033b8af49bcec80ab713eb3dde2a7

File: changelog/803.feature.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (shiftinv) [reply changelog/803.feature.rst]: https://github.com/DisnakeDev/disnake/pull/803/commits/f59f9175b98033b8af49bcec80ab713eb3dde2a7

File: disnake/ui/item.py (hunk 7 above)
Comment (onerandomusername) [reply disnake/ui/item.py:84]: I think alphabetical in all of those instances looks better and is easier to read in documentation.

File: docs/api.rst [hunk 12]
```diff
@@ -6019,10 +6019,6 @@ ActionRow
     :members:
     :exclude-members: add_select
 
-    .. method:: add_select
```
Comment (onerandomusername) [inline docs/api.rst]: this documentation was fine to stay, I think my comment didn't make sense. 

This should stay documented.

File: docs/api.rst (hunk 9 above)
Comment (onerandomusername) [reply docs/api.rst]: I think this should be deprecated but no plans to remove them until a major version bump (or extended time on 2.x)

My future plan is this method can act as a factory for any of the currently four select types.

File: disnake/ui/item.py (hunk 7 above)
Comment (shiftinv) [reply disnake/ui/item.py]: https://github.com/DisnakeDev/disnake/pull/803/commits/16e69da1647e784bdfb25b188b7fbaf1d9af4a96

File: docs/api.rst (hunk 12 above)
Comment (shiftinv) [reply docs/api.rst]: The change I made was independent of your comment, I removed `add_select` again as the other "old" names are also not documented, which is intentional.

**Review by onerandomusername — APPROVED** [review]
